File: simter-id-data/src/main/java/tech/simter/id/generator/impl/TypeIdGenerator.java
package tech.simter.id.generator.impl;

import reactor.core.publisher.Mono;
import tech.simter.id.dao.IdDao;
import tech.simter.id.generator.IdGenerator;

/**
 * An auto increment {@link Long} {@link IdGenerator} base on the specific type.
 * <p>
 * Each type's id is start from 1 and step 1.
 * <p>
 * This {@link IdGenerator} use {@link IdDao#nextLong(String)} to get the next id.
 *
 * @author RJ
 * @see IdDao#nextLong(String)
 */
public class TypeIdGenerator implements IdGenerator<Mono<Long>> {
  private String type;
  private IdDao idDao;

  public TypeIdGenerator(String type, IdDao idDao) {
    this.type = type;
    this.idDao = idDao;
  }

  @Override
  public Mono<Long> nextId() {
    return idDao.nextLong(type);
  }
}

File: simter-id-data/src/main/kotlin/tech/simter/id/generator/impl/TypeIdGenerator.kt
package tech.simter.id.generator.impl

import reactor.core.publisher.Mono
import tech.simter.id.dao.IdDao
import tech.simter.id.IdGenerator

/**
 * An auto increment [Long] [IdGenerator] base on the specific type.
 *
 *
 * Each type's id is start from 1 and step 1.
 *
 *
 * This [IdGenerator] use [IdDao.nextLong] to get the next id.
 *
 * @author RJ
 * @see IdDao.nextLong
 */
class TypeIdGenerator(private val type: String, private val idDao: IdDao) : IdGenerator<Mono<Long>> {

  override fun nextId(): Mono<Long> {
    return idDao.nextLong(type)
  }
}